File: NutzMahn/Arrays & Hashing/238_product_of_array_except_self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;

        int[] sol = new int[n];
        int[] pref = new int[n];
        int[] suff = new int[n];

        pref[0] = 1;
        suff[n-1] = 1;

        for(int i = 1; i < n; i++){
            pref[i] = pref[i - 1] * nums[i - 1];
        }

        for(int i = n - 2; i >= 0; i--){
            suff[i] = suff[i + 1] * nums[i + 1];
        }

        for(int i = 0; i < n; i++){
            sol[i] = pref[i] * suff[i];
        }

        return sol;
    }
}

File: swdevelop24/Arrays_Hashing/238_product_of_array_except_self.py
"""
class Solution:
    def produtExceptSelf(self, nums: list[int]) -> list[int]:
        ret = []
        cal = 1
        for i in range(len(nums)):
            j = 0
            while j < len(nums):
                if j == i:
                    j += 1
                    continue
                cal *= nums[j]
                j += 1
            ret.append(cal)
            cal = 1
        return ret


sol = Solution()
print(sol.produtExceptSelf([1, 2, 3, 4]))
print(sol.produtExceptSelf([-1, 1, 0, -3, 3]))
"""

# time limit error


# different approach
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        n = len(nums)
        prefix = 1
        postfix = 1
        output = [0] * n
        for i in range(n):
            output[i] = prefix
            prefix *= nums[i]
        for i in range(n - 1, -1, -1):
            output[i] *= postfix
            postfix *= nums[i]
        return output


sol = Solution()
print(sol.productExceptSelf([1, 2, 3, 4]))
print(sol.productExceptSelf([-1, 1, 0, -3, 3]))
